src/channels/minesweeper/index.tsx

```diff
@@ -0,0 +1,238 @@
+import type { FormattedDonation, Total } from '@gdq/types/tracker';
+
+import { Face } from './Face';
+import { Tile, TileData } from './Tile';
+import { TILE_DIMENSION, GRID_COLUMNS, GRID_ROWS, TILE_MAP, MINE_CHANCE, mineNumberTiles } from './constants';
+import { createTileCluster, getMineCount, randomFromArray, splitTileIndex } from './utils';
+
+registerChannel('Minesweeper', 132, Minesweeper);
+
+type GridState = {
+	grid: TileData[][];
+	mines: string[];
+	nonMines: string[];
+};
+
+function generateInitialGridState(): GridState {
+	const state: GridState = { grid: [], mines: [], nonMines: [] };
+	for (let i = 0; i < GRID_ROWS; i++) {
+		state.grid[i] = [];
+		for (let j = 0; j < GRID_COLUMNS; j++) {
+			const isMine = Math.random() < MINE_CHANCE;
+			const id = `${i}:${j}`;
+			state.grid[i][j] = {
+				id,
+				isMine,
+				tileType: TILE_MAP.HIDDEN,
+			};
+			if (isMine) {
+				state.mines.push(id);
+			} else {
+				state.nonMines.push(id);
+			}
+		}
+	}
+	return state;
+}
+
+const actions = {
+	RESET: 'reset',
+	FLAG_TILE: 'flag',
+	REVEAL_TILES: 'reveal',
+} as const;
+
+type GridAction =
+	| { type: typeof actions.FLAG_TILE }
+	| { type: typeof actions.RESET }
+	| { type: typeof actions.REVEAL_TILES; revealChance: number };
+
+function gridReducer(state: GridState, action: GridAction) {
+	switch (action.type) {
+		case actions.RESET: {
+			return generateInitialGridState();
+		}
+		case actions.FLAG_TILE: {
+			if (state.mines.length > 0) {
+				const mineIndexStr = randomFromArray(state.mines);
+				const [rowIndex, tileIndex] = splitTileIndex(mineIndexStr);
+
+				const newGrid = state.grid.map((row, ri) => {
+					if (ri !== rowIndex) return row;
+					return row.map<TileData>((tile, ti) =>
+						ti === tileIndex
+							? {
+									...tile,
+									tileType: TILE_MAP.FLAGGED,
+							  }
+							: tile,
+					);
+				});
+
+				const newMines = state.mines.filter((mineIndex) => mineIndex !== mineIndexStr);
+
+				return { ...state, grid: newGrid, mines: newMines };
+			}
+			return state;
+		}
+		case actions.REVEAL_TILES: {
+			if (state.nonMines.length > 0) {
+				const tileId = randomFromArray(state.nonMines);
+				const tiles = createTileCluster(state.grid, tileId, action.revealChance);
+
+				const grid = state.grid.map((row) => {
+					return row.map<TileData>((tile) => {
+						const mineCount = getMineCount(state.grid, tile.id);
+						return tiles.includes(tile?.id)
+							? {
+									...tile,
+									tileType: mineNumberTiles[mineCount],
+							  }
+							: tile;
+					});
+				});
+
+				const nonMines = state.nonMines.filter((id) => !tiles.includes(id));
+
+				return { ...state, grid, nonMines };
+			}
+			return state;
+		}
+		default: {
+			return state;
+		}
+	}
+}
+
+export function Minesweeper(props: ChannelProps) {
+	const [total] = useReplicant<Total | null>('total', null);
+
+	const [gridState, dispatch] = useReducer(gridReducer, undefined, generateInitialGridState);
```

Comment:
Could the grid state possibly be stored in and loaded from a replicant? Similar to how the Breakout and Tetris channels are able to. Because as-is, there isn't really enough time between channel change redemptions to realistically see a board completely fill, so having it be something that can be worked towards over multiple showings would work better IMO.